pkg/operators/wasm/maps.go

```diff
@@ -0,0 +1,208 @@
+			wapi.ValueTypeI64, // Flag
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "mapDelete", i.mapDelete,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // Map
+			wapi.ValueTypeI64, // Key pointer
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+}
+
+// getMap gets an existing map.
+// Params:
+// - stack[0] is the name of the map (string encoded)
+// Return value:
+// - Map handle on success, 0 on error
+func (i *wasmOperatorInstance) getMap(ctx context.Context, m wapi.Module, stack []uint64) {
+	mapNamePtr := stack[0]
+
+	mapName, err := stringFromStack(m, mapNamePtr)
+	if err != nil {
+		i.logger.Warnf("getMap: reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+
+	ebpfMap, ok := i.gadgetCtx.GetVar(operators.MapPrefix + mapName)
+	if !ok {
+		i.logger.Warnf("get map: no map for name %q", mapName)
+		stack[0] = 0
+		return
+	}
+
+	ebpfMap, ok = ebpfMap.(*ebpf.Map)
+	if !ok {
+		i.logger.Warnf("get map: map %q is not an ebpf map", mapName)
+		stack[0] = 0
+		return
+	}
+
+	stack[0] = wapi.EncodeU32(i.addHandle(ebpfMap))
+}
+
+// mapLookup searches the map for a value corresponding to the given key.
+// Params:
+// - stack[0]: Map handle
+// - stack[1]: Key pointer
+// - stack[2]: Value pointer
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) mapLookup(ctx context.Context, m wapi.Module, stack []uint64) {
+	mapHandle := wapi.DecodeU32(stack[0])
+	keyPtr := stack[1]
+	valuePtr := stack[2]
+
+	ebpfMap, ok := getHandle[*ebpf.Map](i, mapHandle)
+	if !ok {
+		stack[0] = 1
+		return
+	}
+
+	key, err := bufFromStack(m, keyPtr)
+	if err != nil {
+		i.logger.Warnf("mapLookup: getting a buf for key pointer: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	value, err := ebpfMap.LookupBytes(key)
+	if err != nil {
+		i.logger.Warnf("mapLookup: getting value: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	err = bufToStack(m, value, valuePtr)
+	if err != nil {
+		i.logger.Warnf("mapLookup: writing back value to stack: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	stack[0] = 0
+}
+
+// mapUpdate updates a value.
+// Params:
+// - stack[0]: Map handle
+// - stack[1]: Key pointer
+// - stack[2]: Value pointer
+// - stack[3]: Flag to update the map
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) mapUpdate(ctx context.Context, m wapi.Module, stack []uint64) {
+	mapHandle := wapi.DecodeU32(stack[0])
+	keyPtr := stack[1]
+	valuePtr := stack[2]
+	flags := stack[3]
+
+	ebpfMap, ok := getHandle[*ebpf.Map](i, mapHandle)
+	if !ok {
+		stack[0] = 1
+		return
+	}
+
+	key, err := bufFromStack(m, keyPtr)
+	if err != nil {
+		i.logger.Warnf("mapLookup: getting a buf for key pointer: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	value, err := bufFromStack(m, valuePtr)
+	if err != nil {
+		i.logger.Warnf("mapLookup: getting a buf for value pointer: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	err = ebpfMap.Update(key, value, ebpf.MapUpdateFlags(flags))
```

Comment:
I think we have to verify the type of ebpf map. For maps of type array or hash, that's fine. But for maps of type tail_call or other maps that accepts a file descriptor as value, there is a risk that the wasm module uses an integer and the kernel interprets that as a fd.

Reply:
For future compatibility, we should verify the content of the flags is among the following values. As of today, the kernel supports:
- BPF_ANY
- BPF_NOEXIST
- BPF_EXIST
- BPF_F_LOCK

And none of those values would cause any problem when used from wasm. And the kernel correctly returns `-EINVAL` when using an unknown flag.

But if the kernel adds a new flag, we should assess whether it could cause a problem for Inspektor Gadget before automatically supporting it. For example, if a future flag causes the addition of a new file descriptor in the process, we would not want to allow it from wasm without more thinking about the correct way to do it.